homeassistant/components/dnsip/config_flow.py

```diff
@@ -0,0 +1,170 @@
+import homeassistant.helpers.config_validation as cv
+
+from .const import (
+    CONF_HOSTNAME,
+    CONF_IPV4,
+    CONF_IPV6,
+    CONF_RESOLVER,
+    CONF_RESOLVER_IPV6,
+    DEFAULT_HOSTNAME,
+    DEFAULT_NAME,
+    DEFAULT_RESOLVER,
+    DEFAULT_RESOLVER_IPV6,
+    DOMAIN,
+)
+
+DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_HOSTNAME, default=DEFAULT_HOSTNAME): cv.string,
+    }
+)
+
+
+async def async_validate_hostname(
+    hostname: str, resolver_ipv4: str, resolver_ipv6: str
+) -> dict[str, bool]:
+    """Validate hostname."""
+
+    async def async_check(hostname: str, resolver: str, qtype: str) -> bool:
+        """Return if able to resolve hostname."""
+        result = False
+        with contextlib.suppress(DNSError):
+            result = bool(
+                await aiodns.DNSResolver(nameservers=[resolver]).query(hostname, qtype)
+            )
+        return result
+
+    result: dict[str, bool] = {}
+
+    tasks = await asyncio.gather(
+        async_check(hostname, resolver_ipv4, "A"),
+        async_check(hostname, resolver_ipv6, "AAAA"),
+    )
+
+    result[CONF_IPV4] = tasks[0]
+    result[CONF_IPV6] = tasks[1]
+
+    return result
+
+
+class DnsIPConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for dnsip integration."""
+
+    VERSION = 1
+
+    @staticmethod
+    @callback
+    def async_get_options_flow(
+        config_entry: config_entries.ConfigEntry,
+    ) -> DnsIPOptionsFlowHandler:
+        """Return Option handler."""
+        return DnsIPOptionsFlowHandler(config_entry)
+
+    async def async_step_import(self, config: dict[str, Any]) -> FlowResult:
+        """Import a configuration from config.yaml."""
+
+        hostname = config.get(CONF_HOSTNAME, DEFAULT_HOSTNAME)
+        self._async_abort_entries_match({CONF_HOSTNAME: hostname})
+        config[CONF_HOSTNAME] = hostname
+        return await self.async_step_user(user_input=config)
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle the initial step."""
+
+        errors = {}
+
+        if user_input:
+
+            hostname = user_input[CONF_HOSTNAME]
+            name = DEFAULT_NAME if hostname == DEFAULT_HOSTNAME else hostname
+            resolver = DEFAULT_RESOLVER
+            resolver_ipv6 = DEFAULT_RESOLVER_IPV6
+
+            validate = await async_validate_hostname(hostname, resolver, resolver_ipv6)
+
+            if not validate[CONF_IPV4] and not validate[CONF_IPV6]:
+                errors["base"] = "invalid_hostname"
+            else:
+                await self.async_set_unique_id(hostname)
+                self._abort_if_unique_id_configured()
+
+                return self.async_create_entry(
+                    title=name,
+                    data={
+                        CONF_HOSTNAME: hostname,
+                        CONF_NAME: name,
+                        CONF_RESOLVER: resolver,
```

Comment:
We can set the default resolvers in the config entry options here directly instead of storing them in the config entry data.

https://github.com/home-assistant/core/blob/dd5bcafab7f68ac726f199af32bd63ed25e37919/homeassistant/config_entries.py#L1438-L1447

Reply:
Would be good to address this before the release tomorrow to avoid needing migration code.

Reply:
I'll ping you in the new PR coming shortly